Fix Virtual Network acceptance tests for OpenNebula 6.4

```
resource_opennebula_virtual_network_test.go:20: Step 1/4 error: Error running apply: exit status 1
[50](https://github.com/OpenNebula/terraform-provider-opennebula/runs/6391905591?check_suite_focus=true#step:10:50)
        
[51](https://github.com/OpenNebula/terraform-provider-opennebula/runs/6391905591?check_suite_focus=true#step:10:51)
        Error: Error: OpenNebula error [ACTION]: [one.vn.add_ar] Could not execute one.vn.add_arVirtual Network is in wrong state: LOCK_CREATE
[52](https://github.com/OpenNebula/terraform-provider-opennebula/runs/6391905591?check_suite_focus=true#step:10:52)
        AR: AR=[
[53](https://github.com/OpenNebula/terraform-provider-opennebula/runs/6391905591?check_suite_focus=true#step:10:53)
            SIZE="12",
[54](https://github.com/OpenNebula/terraform-provider-opennebula/runs/6391905591?check_suite_focus=true#step:10:54)
            TYPE="IP4",
[55](https://github.com/OpenNebula/terraform-provider-opennebula/runs/6391905591?check_suite_focus=true#step:10:55)
            IP="172.16.100.130" ]
[56](https://github.com/OpenNebula/terraform-provider-opennebula/runs/6391905591?check_suite_focus=true#step:10:56)
        
[57](https://github.com/OpenNebula/terraform-provider-opennebula/runs/6391905591?check_suite_focus=true#step:10:57)
          with opennebula_virtual_network.test,
[58](https://github.com/OpenNebula/terraform-provider-opennebula/runs/6391905591?check_suite_focus=true#step:10:58)
          on terraform_plugin_test.tf line 2, in resource "opennebula_virtual_network" "test":
[59](https://github.com/OpenNebula/terraform-provider-opennebula/runs/6391905591?check_suite_focus=true#step:10:59)
           2: resource "opennebula_virtual_network" "test" {
[60](https://github.com/OpenNebula/terraform-provider-opennebula/runs/6391905591?check_suite_focus=true#step:10:60)
        
[61](https://github.com/OpenNebula/terraform-provider-opennebula/runs/6391905591?check_suite_focus=true#step:10:61)
    testing_new.go:70: Error running post-test destroy, there may be dangling resources: Expected virtual network 4 to have been destroyed
[62](https://github.com/OpenNebula/terraform-provider-opennebula/runs/6391905591?check_suite_focus=true#step:10:62)
--- FAIL: TestAccVirtualNetwork (0.66s)
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,6 +4,7 @@ NOTES:
 
 * All datasources schemas have been reworked and an independant read method has been added for each.
 * The provider has been migrated to use the SDK v2.
+* OpenNebula binding (goca) dependency has been updated to the 6.4 release.
 
 FEATURES:
 
@@ -20,6 +21,7 @@ ENHANCEMENTS:
 * data/opennebula_virtual_data_center: make `name` optional and add `tags` filtering
 * data/opennebula_virtual_network: make `name` optional and enable `tags` filtering
 * resources/opennebula_group: add `sunstone` and `tags` sections
+* resources/opennebula_virtual_network: compatibility added for network states
 
 FEATURES:
 
@@ -39,6 +41,8 @@ BUG FIXES:
 
 * resources/opennebula_security_group: read `name`
 * resources/opennebula_virtual_machine: fix volatile disk update adding `computed_volatile_format`
+* resources/opennebula_virtual_machine: fix template quotes escaping
+* resources/opennebula_template: fix template quotes escaping
 
 ## 0.4.3 (March 23th, 2022)
 

--- a/go.mod
+++ b/go.mod
@@ -3,7 +3,8 @@ module github.com/OpenNebula/terraform-provider-opennebula
 go 1.13
 
 require (
-	github.com/OpenNebula/one/src/oca/go/src/goca v0.0.0-20210824074935-e55e3a53c4c0
+	github.com/OpenNebula/one/src/oca/go/src/goca v0.0.0-20220509171820-bf6ef68bd15a
+	github.com/hashicorp/go-version v1.5.0 // indirect
 	github.com/hashicorp/terraform-plugin-sdk/v2 v2.10.1
 	github.com/kolo/xmlrpc v0.0.0-20190909154602-56d5ec7c422e // indirect
 )

--- a/opennebula/resource_opennebula_template.go
+++ b/opennebula/resource_opennebula_template.go
@@ -9,7 +9,6 @@ import (
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 
 	"github.com/OpenNebula/one/src/oca/go/src/goca"
-	dyn "github.com/OpenNebula/one/src/oca/go/src/goca/dynamic"
 	"github.com/OpenNebula/one/src/oca/go/src/goca/parameters"
 	"github.com/OpenNebula/one/src/oca/go/src/goca/schemas/shared"
 	"github.com/OpenNebula/one/src/oca/go/src/goca/schemas/vm"
@@ -580,26 +579,7 @@ func resourceOpennebulaTemplateUpdate(d *schema.ResourceData, meta interface{}) 
 	}
 
 	update := false
-	newTpl := vm.NewTemplate()
-
-	// copy existing template and re-escape chars if needed
-	for _, e := range tpl.Template.Elements {
-		pair, ok := e.(*dyn.Pair)
-		if ok {
-			// copy a pair
-			escapedValue := strings.ReplaceAll(pair.Value, "\"", "\\\"")
-			newTpl.AddPair(e.Key(), escapedValue)
-		} else {
-			// copy a vector
-			vector, _ := e.(*dyn.Vector)
-
-			newVec := newTpl.AddVector(e.Key())
-			for _, p := range vector.Pairs {
-				escapedValue := strings.ReplaceAll(p.Value, "\"", "\\\"")
-				newVec.AddPair(p.Key(), escapedValue)
-			}
-		}
-	}
+	newTpl := tpl.Template
 
 	if d.HasChange("features") {
 		newTpl.Del("FEATURES")
@@ -639,8 +619,7 @@ func resourceOpennebulaTemplateUpdate(d *schema.ResourceData, meta interface{}) 
 		schedRequirements := d.Get("sched_requirements").(string)
 
 		if len(schedRequirements) > 0 {
-			escapedValue := strings.ReplaceAll(schedRequirements, "\"", "\\\"")
-			newTpl.Placement(vmk.SchedRequirements, escapedValue)
+			newTpl.Placement(vmk.SchedRequirements, schedRequirements)
 		} else {
 			newTpl.Del(string(vmk.SchedRequirements))
 		}
@@ -651,8 +630,7 @@ func resourceOpennebulaTemplateUpdate(d *schema.ResourceData, meta interface{}) 
 		schedDSRequirements := d.Get("sched_ds_requirements").(string)
 
 		if len(schedDSRequirements) > 0 {
-			escapedValue := strings.ReplaceAll(schedDSRequirements, "\"", "\\\"")
-			newTpl.Placement(vmk.SchedDSRequirements, escapedValue)
+			newTpl.Placement(vmk.SchedDSRequirements, schedDSRequirements)
 		} else {
 			newTpl.Del(string(vmk.SchedDSRequirements))
 		}

--- a/opennebula/resource_opennebula_template_test.go
+++ b/opennebula/resource_opennebula_template_test.go
@@ -2,11 +2,12 @@ package opennebula
 
 import (
 	"fmt"
-	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
-	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
 	"reflect"
 	"strconv"
 	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
 
 	"github.com/OpenNebula/one/src/oca/go/src/goca"
 	"github.com/OpenNebula/one/src/oca/go/src/goca/schemas/shared"
@@ -110,7 +111,7 @@ func TestAccTemplate(t *testing.T) {
 						"virtio_scsi_queues": "1",
 						"acpi":               "YES",
 					}),
-					resource.TestCheckResourceAttr("opennebula_template.template", "sched_requirements", "CLUSTER_ID!=\\\"123\\\""),
+					resource.TestCheckResourceAttr("opennebula_template.template", "sched_requirements", "CLUSTER_ID!=\"123\""),
 					resource.TestCheckResourceAttr("opennebula_template.template", "description", "Template created for provider acceptance tests - updated"),
 					resource.TestCheckResourceAttrSet("opennebula_template.template", "uid"),
 					resource.TestCheckResourceAttrSet("opennebula_template.template", "gid"),
@@ -337,7 +338,7 @@ resource "opennebula_template" "template" {
     version = "2"
   }
 
-  sched_requirements = "CLUSTER_ID!=\\\"123\\\""
+  sched_requirements = "CLUSTER_ID!=\"123\""
 
 }
 `

--- a/opennebula/resource_opennebula_virtual_machine.go
+++ b/opennebula/resource_opennebula_virtual_machine.go
@@ -1021,33 +1021,17 @@ func resourceOpennebulaVirtualMachineUpdate(d *schema.ResourceData, meta interfa
 	}
 
 	update := false
-	tpl := vm.NewTemplate()
-
-	// copy existing template and re-escape chars if needed
-	for _, e := range vmInfos.UserTemplate.Template.Elements {
-		pair, ok := e.(*dyn.Pair)
-		if ok {
-			// copy a pair
-			escapedValue := strings.ReplaceAll(pair.Value, "\"", "\\\"")
-			tpl.AddPair(e.Key(), escapedValue)
-		} else {
-			// copy a vector
-			vector, _ := e.(*dyn.Vector)
-
-			newVec := tpl.AddVector(e.Key())
-			for _, p := range vector.Pairs {
-				escapedValue := strings.ReplaceAll(p.Value, "\"", "\\\"")
-				newVec.AddPair(p.Key(), escapedValue)
-			}
-		}
+	tpl := &vm.Template{
+		Template: dyn.Template{
+			Elements: vmInfos.UserTemplate.Template.Elements,
+		},
 	}
 
 	if d.HasChange("sched_requirements") {
 		schedRequirements := d.Get("sched_requirements").(string)
 
 		if len(schedRequirements) > 0 {
-			escapedValue := strings.ReplaceAll(schedRequirements, "\"", "\\\"")
-			tpl.Placement(vmk.SchedRequirements, escapedValue)
+			tpl.Placement(vmk.SchedRequirements, schedRequirements)
 		} else {
 			tpl.Del(string(vmk.SchedRequirements))
 		}
@@ -1058,8 +1042,7 @@ func resourceOpennebulaVirtualMachineUpdate(d *schema.ResourceData, meta interfa
 		schedDSRequirements := d.Get("sched_ds_requirements").(string)
 
 		if len(schedDSRequirements) > 0 {
-			escapedValue := strings.ReplaceAll(schedDSRequirements, "\"", "\\\"")
-			tpl.Placement(vmk.SchedDSRequirements, escapedValue)
+			tpl.Placement(vmk.SchedDSRequirements, schedDSRequirements)
 		} else {
 			tpl.Del(string(vmk.SchedDSRequirements))
 		}

--- a/opennebula/resource_opennebula_virtual_machine_test.go
+++ b/opennebula/resource_opennebula_virtual_machine_test.go
@@ -160,7 +160,7 @@ func TestAccVirtualMachine(t *testing.T) {
 					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "disk.#", "1"),
 					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "tags.%", "1"),
 					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "tags.env", "dev"),
-					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "sched_requirements", "CLUSTER_ID!=\\\"123\\\""),
+					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "sched_requirements", "CLUSTER_ID!=\"123\""),
 					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "timeout", "4"),
 					resource.TestCheckResourceAttrSet("opennebula_virtual_machine.test", "uid"),
 					resource.TestCheckResourceAttrSet("opennebula_virtual_machine.test", "gid"),
@@ -200,7 +200,7 @@ func TestAccVirtualMachine(t *testing.T) {
 					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "tags.customer", "test"),
 					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "tags.version", "3"),
 					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "description", "This is an acceptance test VM"),
-					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "sched_requirements", "CLUSTER_ID!=\\\"123\\\""),
+					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "sched_requirements", "CLUSTER_ID!=\"123\""),
 					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "timeout", "4"),
 					resource.TestCheckResourceAttrSet("opennebula_virtual_machine.test", "uid"),
 					resource.TestCheckResourceAttrSet("opennebula_virtual_machine.test", "gid"),
@@ -239,7 +239,7 @@ func TestAccVirtualMachine(t *testing.T) {
 					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "tags.env", "dev"),
 					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "tags.customer", "test2"),
 					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "description", "VM created for provider acceptance tests"),
-					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "sched_requirements", "CLUSTER_ID!=\\\"123\\\""),
+					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "sched_requirements", "CLUSTER_ID!=\"123\""),
 					resource.TestCheckResourceAttr("opennebula_virtual_machine.test", "timeout", "4"),
 					resource.TestCheckResourceAttrSet("opennebula_virtual_machine.test", "uid"),
 					resource.TestCheckResourceAttrSet("opennebula_virtual_machine.test", "gid"),
@@ -859,7 +859,7 @@ resource "opennebula_virtual_machine" "test" {
     env = "dev"
   }
 
-  sched_requirements = "CLUSTER_ID!=\\\"123\\\""
+  sched_requirements = "CLUSTER_ID!=\"123\""
 
   timeout = 4
 }
@@ -899,7 +899,7 @@ resource "opennebula_virtual_machine" "test" {
     version = "3"
   }
 
-  sched_requirements = "CLUSTER_ID!=\\\"123\\\""
+  sched_requirements = "CLUSTER_ID!=\"123\""
 
   timeout = 4
 }
@@ -938,7 +938,7 @@ resource "opennebula_virtual_machine" "test" {
     customer = "test2"
   }
 
-  sched_requirements = "CLUSTER_ID!=\\\"123\\\""
+  sched_requirements = "CLUSTER_ID!=\"123\""
 
   timeout = 4
 }

--- a/opennebula/resource_opennebula_virtual_network.go
+++ b/opennebula/resource_opennebula_virtual_network.go
@@ -6,7 +6,10 @@ import (
 	"net"
 	"strconv"
 	"strings"
+	"time"
 
+	"github.com/hashicorp/go-version"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 
 	"github.com/OpenNebula/one/src/oca/go/src/goca"
@@ -23,6 +26,9 @@ func resourceOpennebulaVirtualNetwork() *schema.Resource {
 		Exists: resourceOpennebulaVirtualNetworkExists,
 		Update: resourceOpennebulaVirtualNetworkUpdate,
 		Delete: resourceOpennebulaVirtualNetworkDelete,
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(3 * time.Minute),
+		},
 		Importer: &schema.ResourceImporter{
 			State: schema.ImportStatePassthrough,
 		},
@@ -439,6 +445,23 @@ func resourceOpennebulaVirtualNetworkCreate(d *schema.ResourceData, meta interfa
 			return err
 		}
 		vnc = controller.VirtualNetwork(vnetID)
+
+		// virtual network states were introduce with OpenNebula 6.4 release
+		ver, err := controller.SystemVersion()
+		if err != nil {
+			return nil
+		}
+
+		oneVersion, err := version.NewVersion(ver)
+		requiredVersion, err := version.NewVersion("6.4.0")
+
+		if oneVersion.GreaterThan(requiredVersion) {
+			timeout := d.Timeout(schema.TimeoutCreate)
+			_, err = waitForVNetworkState(vnc, timeout, "READY")
+			if err != nil {
+				return fmt.Errorf("Error waiting for Image (%s) to be in state READY: %s", d.Id(), err)
+			}
+		}
 
 		d.SetId(fmt.Sprintf("%v", vnetID))
 
@@ -1176,4 +1199,45 @@ func resourceOpennebulaVirtualNetworkDelete(d *schema.ResourceData, meta interfa
 
 	log.Printf("[INFO] Successfully deleted Vnet\n")
 	return nil
+}
+
+func waitForVNetworkState(vnc *goca.VirtualNetworkController, timeout time.Duration, state ...string) (interface{}, error) {
+
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{"anythingelse"},
+		Target:  state,
+		Refresh: func() (interface{}, string, error) {
+
+			log.Println("Refreshing virtual network state...")
+
+			var vNetInfos *vn.VirtualNetwork
+			vNetInfos, err := vnc.Info(false)
+			if err != nil {
+				if NoExists(err) {
+					return vNetInfos, "notfound", nil
+				}
+				return vNetInfos, "", err
+			}
+			state, err := vNetInfos.State()
+			if err != nil {
+				return vNetInfos, "", err
+			}
+
+			log.Printf("Image (ID:%d, name:%s) is currently in state %v", vNetInfos.ID, vNetInfos.Name, state.String())
+
+			switch state {
+			case vn.Ready, vn.LockCreate, vn.LockDelete:
+				return vNetInfos, state.String(), nil
+			case vn.Error:
+				return vNetInfos, state.String(), fmt.Errorf("Image (ID:%d) entered error state.", vNetInfos.ID)
+			default:
+				return vNetInfos, "anythingelse", nil
+			}
+		},
+		Timeout:    timeout,
+		Delay:      10 * time.Second,
+		MinTimeout: 3 * time.Second,
+	}
+
+	return stateConf.WaitForState()
 }

--- a/opennebula/resource_opennebula_virtual_network.go
+++ b/opennebula/resource_opennebula_virtual_network.go
@@ -459,7 +459,7 @@ func resourceOpennebulaVirtualNetworkCreate(d *schema.ResourceData, meta interfa
 			timeout := d.Timeout(schema.TimeoutCreate)
 			_, err = waitForVNetworkState(vnc, timeout, "READY")
 			if err != nil {
-				return fmt.Errorf("Error waiting for Image (%s) to be in state READY: %s", d.Id(), err)
+				return fmt.Errorf("Error waiting for virtual network (ID:%s) to be in state READY: %s", d.Id(), err)
 			}
 		}
 
@@ -1223,13 +1223,13 @@ func waitForVNetworkState(vnc *goca.VirtualNetworkController, timeout time.Durat
 				return vNetInfos, "", err
 			}
 
-			log.Printf("Image (ID:%d, name:%s) is currently in state %v", vNetInfos.ID, vNetInfos.Name, state.String())
+			log.Printf("virtual network (ID:%d, name:%s) is currently in state %v", vNetInfos.ID, vNetInfos.Name, state.String())
 
 			switch state {
 			case vn.Ready, vn.LockCreate, vn.LockDelete:
 				return vNetInfos, state.String(), nil
 			case vn.Error:
-				return vNetInfos, state.String(), fmt.Errorf("Image (ID:%d) entered error state.", vNetInfos.ID)
+				return vNetInfos, state.String(), fmt.Errorf("virtual network (ID:%d) entered error state.", vNetInfos.ID)
 			default:
 				return vNetInfos, "anythingelse", nil
 			}

--- a/opennebula/shared_schemas.go
+++ b/opennebula/shared_schemas.go
@@ -499,16 +499,13 @@ func generateVMTemplate(d *schema.ResourceData, tpl *vm.Template) error {
 
 	schedReq, ok := d.GetOk("sched_requirements")
 	if ok {
-		schedReqStr := strings.ReplaceAll(schedReq.(string), "\"", "\\\"")
-
-		tpl.AddPair("SCHED_REQUIREMENTS", schedReqStr)
+		tpl.AddPair("SCHED_REQUIREMENTS", schedReq.(string))
 
 	}
 
 	schedDSReq, ok := d.GetOk("sched_ds_requirements")
 	if ok {
-		schedDSReqStr := strings.ReplaceAll(schedDSReq.(string), "\"", "\\\"")
-		tpl.AddPair("SCHED_DS_REQUIREMENTS", schedDSReqStr)
+		tpl.AddPair("SCHED_DS_REQUIREMENTS", schedDSReq.(string))
 
 	}
 

